This fixes the Capacity card logic

aruniiird opened the pull request:
This PR is to correct the logic issue with the noobaa capacity card logic.

Fix Issue: https://github.com/openshift/console/issues/2145

Signed-off-by: aruniiird <arun.iiird@gmail.com>

aruniiird commented:
I have rephrased the capacity usage card logic (according to the meeting with NooBaa developers)... @cloudbehl , @rawagner , @afreen23 , can you guys take look at the code...
Here is the [NooBaa doc](https://docs.google.com/document/d/15lugKm-ljNIJiDRfKBhnHWNwdqrF04dPw5cgDZ93-UA/edit#heading=h.jxj6d3pghiuz)

afreen23 commented:
@aruniiird can you add a screenshot with the prometheus data that verifies , this is correct

cloudbehl commented:
/hold

aruniiird commented:
Fixed the UI issues ( thanks to @afreen23  ) and corrected the logic according to the new sorted query.

with data

![capacity_usage_with_data](https://user-images.githubusercontent.com/895234/62349406-4ce0f100-b51d-11e9-9121-018d6e5dae2c.png)

with only 'Others' label and ZERO value

![capacity_usage_with_empty_others_data](https://user-images.githubusercontent.com/895234/62349454-700ba080-b51d-11e9-858e-e11e07fe7777.png)

@cloudbehl , @rawagner , @afreen23  please review

rawagner commented:
> with only 'Others' label and ZERO value
> 
> ![capacity_usage_with_empty_others_data](https://user-images.githubusercontent.com/895234/62349454-700ba080-b51d-11e9-858e-e11e07fe7777.png)

This looks weird, cant we show full blue circle instead ?

rawagner commented:
> > with only 'Others' label and ZERO value
> > ![capacity_usage_with_empty_others_data](https://user-images.githubusercontent.com/895234/62349454-700ba080-b51d-11e9-858e-e11e07fe7777.png)
> 
> This looks weird, cant we show full blue circle instead ?

@andybraren

andybraren commented:
I agree @rawagner, a full blue circle is probably best (and hopefully easiest) for this zero state.

@yuvalgalanti seeing “Capacity Usage” in the title and then “Usage Breakdown” right below it seems odd. What do you think about renaming this card something like “Data Breakdown” (to align with the other cards that refer to “Data”) and then getting rid of the second heading? Would that make sense?

aruniiird commented:
> > with only 'Others' label and ZERO value
> > ![capacity_usage_with_empty_others_data](https://user-images.githubusercontent.com/895234/62349454-700ba080-b51d-11e9-858e-e11e07fe7777.png)
> 
> This looks weird, cant we show full blue circle instead ?

This effect (ie; showing full blue circle) could easily be attained by the following [line](https://github.com/openshift/console/pull/2146/files#diff-bed24a346415a46422a4b6761843e74bR35) change in 'capacity-card-body.tsx', 

```suggestion
    m.y = total ? (m.y / total) * 100 : 100;
```
![full_blue_circle_zero_others](https://user-images.githubusercontent.com/895234/62414773-d25bc280-b63d-11e9-8b60-a25ebe38f954.png)

But this goes a bit weird when we have multiple items (other than the label 'Others') which are not using any capacity (ie; ZERO capacity usage).

![multiple_zero_valued_labels](https://user-images.githubusercontent.com/895234/62414959-07691480-b640-11e9-8d94-f68a4fa89d81.png)

This scenario (ie; multiple items with ZERO capacity usage) may not occur at all, but if we have a confirmation I could go with this minimal change.

aruniiird commented:
@rawagner , @andybraren , Sorry, I missed to tag you in my reply... Can you please check the above comment, and let me know your suggestions...

andybraren commented:
I think I'm okay with the new images you posted @aruniiird. I agree it's a bit odd, but less odd than seeing no circle at all and the floating `0 i`.

One last thing from me - can we change `0 i` to `0 Mi` when in this ZERO state? `i` isn't a real unit. `0 Mi` should hopefully make things a little clearer.

(Side note: We're currently figuring out if we can change `Mi/Gi/Ti` to `MiB/GiB/TiB` across all unit labels in our dashboards and the rest of the OpenShift Console, but haven't made a decision on that yet).

afreen23 commented on frontend/packages/noobaa-storage-plugin/src/components/capacity-card/capacity-card-body.tsx:
```diff
@@ -1,72 +1,52 @@
+  // get the percentage
+  metricData.forEach((m) => {
+    metricLegend.push({ name: `${m.x}` });
+    m.y = total ? (m.y / total) * 100 : 0;
```
some suitable `humanize` function should be used here if possible. The `(m.y / total) * 100` might give some crazy value like `14.23344`

afreen23 commented on frontend/packages/noobaa-storage-plugin/src/components/capacity-card/capacity-card-body.tsx:
```diff
@@ -1,72 +1,52 @@
   }
   // trying to avoid redundant code
-  const { metricData, metricLegend, total } = getMetricDataAndLegend(metricsData);
+  const { metricData, metricLegend } = getMetricDataAndLegend(metricsData, totalUsage);
```
I see the comment does not help much here

aruniiird replied:
Removed...

afreen23 commented on frontend/packages/noobaa-storage-plugin/src/components/capacity-card/capacity-card-body.tsx:
```diff
@@ -76,31 +56,30 @@ export const CapacityCardBody: React.FC<CapacityCardBodyProps> = ({ isLoading, m
-        radius={53}
-        innerRadius={46}
-      />
+    <div className="nb-capacity-card__body">
```
can we rename it to `nb-capacity-card__chart-body` to not confuse with the css for`<CapacityCardBody>`

aruniiird replied:
Done...

afreen23 commented on frontend/packages/noobaa-storage-plugin/src/components/capacity-card/capacity-card-body.tsx:
```diff
@@ -76,31 +56,30 @@ export const CapacityCardBody: React.FC<CapacityCardBodyProps> = ({ isLoading, m
+          legendData={metricLegend}
+          legendOrientation="vertical"
+          data={metricData}
+          labels={(datum) => `${datum.x}: ${datum.y.toFixed(2)}%`}
```
I think here we can use `humanizePercentage` in `units.js` .

aruniiird replied:
I probably have to use some thing like below
```typescript
           `${datum.x}: ${humanizePercentage(totalUsage ? datum.y.toFixed(1) : 0).string}`
```
instead of
``` typescript
          `${datum.x}: ${totalUsage ? datum.y.toFixed(1) : 0}%`
```
If you insist, will change it, but I prefer to go with a simple solution...

rawagner replied:
its better to use `humanizePercentage` because of consistency. You are using 2 fraction digits, but rest of the console is using `humanizePercentage` which uses 1 fraction digit. And once we change that it will be consistently changed everywhere.

i think that simple
```
${datum.x}: ${humanizePercentage(datum.y).string}
```
should work

aruniiird replied:
@afreen23 had the same point... Will change it...

aruniiird replied:
Updated the PR

afreen23 commented on frontend/packages/noobaa-storage-plugin/src/components/capacity-card/capacity-card-body.tsx:
```diff
@@ -1,72 +1,52 @@
+  // get the percentage
+  metricData.forEach((m) => {
+    metricLegend.push({ name: `${m.x}` });
+    m.y = total ? (m.y / total) * 100 : 0;
```
The value coming from prometheus is in bytes.
It needs to be converted into proper unit before using in chart. Not doing this will lead to discrepancy while showing **labels** . 
Try using `humanizeBinaryBytesWithoutB` for `m.y` and `total` (ensure that both the units be same)

afreen23 replied:
![Screenshot from 2019-08-06 17-04-27](https://user-images.githubusercontent.com/25664409/62536775-77f47900-b86c-11e9-8f9a-ad973c5c2754.png)

aruniiird replied:
> It needs to be converted into proper unit before using in chart. Not doing this will lead to discrepancy while showing labels
This should never happen, as we are showing the percentage. In the above graph, the data given has anomaly. The total value (the number shown in the middle of the chart) should always be higher than any of the individual value.
As we are calculating percentage, we don't have to do any conversions, but have to make sure that all the data arriving are of same unit...

afreen23 replied:
Okay I have checked , the mocked data was incorrect.

cloudbehl commented:
/ok-to-test

cloudbehl commented:
@aruniiird remove the lint errors.

aruniiird commented on frontend/packages/noobaa-storage-plugin/src/components/capacity-card/capacity-card-body.scss:
```diff
@@ -0,0 +1,8 @@
+.nb-capacity-card__title {
+  font-weight: 100;
```
May the name (`.nb-capacity-card__title`) is confusing everyone. This 'style' is being used to show a label (`Usage Breakdown`). Will rename the scss style to `.nb-capacity-card__label`.

afreen23 replied:
okay I see, @andybraren could you please confirm if `Usage Breakdown` is a title or not ?
https://invis.io/BRS4A5ETN3H#/364134659_Object_storage_Dashboard-_IO_operations_for_providers

afreen23 replied:
oh I see its resolved now, thanks!

aruniiird commented on frontend/packages/noobaa-storage-plugin/src/components/capacity-card/capacity-card-body.scss:
```diff
@@ -0,0 +1,8 @@
+  font-weight: 100;
+  font-size: 1rem;
+}
+.nb-capacity-card__body {
```
Acknowledged...

aruniiird commented on frontend/packages/noobaa-storage-plugin/src/components/capacity-card/capacity-card.tsx:
```diff
@@ -11,22 +11,22 @@ import {
+import { PrometheusResponse } from '@console/internal/components/graphs';
+import { getInstantVectorStats } from '@console/internal/components/graphs/utils';
 import { CapacityCardBody } from './capacity-card-body';
+import { ObjectDashboardQuery, ObjectCapacityQueries } from '../../queries';
```
Acknowledged...

aruniiird commented on frontend/packages/noobaa-storage-plugin/src/components/capacity-card/capacity-card.tsx:
```diff
@@ -36,41 +36,58 @@ const ObjectDashboardCapacityCard: React.FC<DashboardItemProps> = ({
   );
+  const capacityUsageTotalStats = getInstantVectorStats(capacityUsageTotalResult, '');
+  console.log('CapacityTotalResult:', capacityUsageTotalResult);
+  console.log('Selected CapacityQueries:', ObjectCapacityQueries[capacityUsageType]);
```
My bad... Deleting the debug lines... (=

aruniiird commented on frontend/packages/noobaa-storage-plugin/src/components/capacity-card/capacity-card.tsx:
```diff
@@ -11,22 +11,22 @@ import {
-export const DataToQueryMap = {
-  Projects: 'project',
-  'Bucket Class': 'bucket_class',
+const dataToQueryMap = {
```
Agreed, will rename the 'var'... 
Initial days, this was used to fetch the correct query, but now we use a different variable...

aruniiird commented on frontend/packages/noobaa-storage-plugin/src/components/capacity-card/capacity-card-body.tsx:
```diff
@@ -1,72 +1,52 @@
+//     with 'Others' label appended to it
 const getMetricDataAndLegend = (
-  metricDataPoint: DataPoint[],
+  metricData: DataPoint[],
```
Done...

aruniiird commented on frontend/packages/noobaa-storage-plugin/src/components/capacity-card/capacity-card-body.tsx:
```diff
@@ -1,72 +1,52 @@
   maxSize: number = 6,
 ): {
-  metricData: MetricDataType[];
+  metricData: DataPoint[];
```
This was used to control number of items to be fetched, but now this is directly done through the query... So not required...

aruniiird commented:
> One last thing from me - can we change `0 i` to `0 Mi` when in this ZERO state? `i` isn't a real unit. `0 Mi` should hopefully make things a little clearer.
 
@andybraren , yup, I agree
Acknowledged...

aruniiird commented:
Some screen shots after adding the review changes...

Zero data image

![cu_zero_only_others](https://user-images.githubusercontent.com/895234/62617783-4a710380-b930-11e9-9ec4-ce729f422840.png)

Normal data image

![cu_with_data](https://user-images.githubusercontent.com/895234/62617790-5066e480-b930-11e9-9369-4a65c5e7852b.png)

aruniiird commented:
@afreen23 , @cloudbehl , @rawagner , @andybraren , let me know if any changes has to be made...

Thanks,
Arun

andybraren commented:
Thank you for the screenshots @aruniiird, it's looking great! I just talked to @yuvalgalanti and we have one last tweak. 🙂

Could we change the title of this card from "Capacity Usage" to "Capacity Breakdown" and then remove the "Usage Breakdown" label that's currently above the chart? This would make the card's purpose clearer, and with fewer words.

aruniiird commented:
> Could we change the title of this card from "Capacity Usage" to "Capacity Breakdown" and then remove the "Usage Breakdown" label that's currently above the chart? This would make the card's purpose clearer, and with fewer words.

@andybraren , pushed the changes...

![capacity_breakdown_card](https://user-images.githubusercontent.com/895234/62668708-7204b200-b9aa-11e9-97a8-a4a867cc9362.png)

andybraren commented:
@aruniiird Awesome, thank you! 👍

FYI @yuvalgalanti

rawagner commented on frontend/packages/noobaa-storage-plugin/src/components/capacity-card/capacity-card-body.tsx:
```diff
@@ -75,32 +53,31 @@ export const CapacityCardBody: React.FC<CapacityCardBodyProps> = ({ isLoading, m
       </EmptyState>
     );
   }
+  const totalHumanized = humanizeBinaryBytesWithoutB(totalUsage, 'Mi');
```
i think this is not correct. `humanizeBinaryBytesWithoutB` arguments are `value`, `initialUnit`, `preferredUnit`. You are setting `initialValue` to `Mi` but it actually is `i`. Correct usage should be  `const totalHumanized = humanizeBinaryBytesWithoutB(totalUsage, null,  totalUsage ? null : 'Mi');` - you want `Mi` as preferred unit only if totalUsage is 0, and initialUnit is always a default one (bytes).

aruniiird replied:
thanks @rawagner ... tested with the mock data, works perfectly fine... (=

aruniiird replied:
Updated the PR...

yuvalgalanti commented:
> @aruniiird

thanks @aruniiird

cloudbehl commented:
/test e2e-aws-console-olm

aruniiird commented:
@rawagner , I have made the said changes... 
@cloudbehl , if no other changes required, this PR should be good to go, can you please remove the 'hold' on this.

rawagner commented:
/approve
/assign @cloudbehl

cloudbehl commented on frontend/packages/noobaa-storage-plugin/src/components/capacity-card/capacity-card-body.tsx:
```diff
@@ -75,32 +57,31 @@ export const CapacityCardBody: React.FC<CapacityCardBodyProps> = ({ isLoading, m
       </EmptyState>
     );
   }
+  const totalHumanized = humanizeBinaryBytesWithoutB(totalUsage, null, totalUsage ? null : 'Mi');
```
Let use humanizeBinaryBytes where ever we are using it now.
```suggestion
  const totalHumanized = humanizeBinaryBytes(totalUsage, null, totalUsage ? null : 'Mi');
```

aruniiird replied:
Just an FYI, 
`humanizeBinaryBytes` are backed by `binaryBytes` data structure and that looks something like this,
```javascript
  binaryBytes: {
    units: ['B', 'KiB', 'MiB', 'GiB', 'TiB', 'PiB'],
    space: true,
    divisor: 1024,
  },
```
The `units` array don't have 'EiB' in it (ideally it should have, in order to be consistent with other data structures)

cloudbehl replied:
@aruniiird Thanks, I think you can send a separate PR to fix this.

aruniiird replied:
@cloudbehl , made the changes... Some screen shot,

![cu_card_with_humanizeBinaryBytes](https://user-images.githubusercontent.com/895234/62773614-365c0c00-bac0-11e9-82e9-fe0995c744d7.png)

PS : For large values (mostly corner cases) we may see an odd behavior with the total chart-title (text in the middle).

rawagner replied:
hm, you're right about those large values. How about moving units below the number ? @yuvalgalanti

andybraren replied:
`font-size: 1.2rem;` seems to work well up to `88.88 GiB` and seems to look better in general, maybe we could do that?

I know we've also talked about rounding to the nearest 1 decimal place ([see here](https://docs.google.com/a/redhat.com/document/d/1oT2n1S7m9wR9Fg8MmQogvLXyRiBgUS_yNFhE0bszMVw/edit?disco=AAAADPPwrHE)) but that's not confirmed yet.

aruniiird replied:
> hm, you're right about those large values. How about moving units below the number ? @yuvalgalanti

Already tried that with my limited knowledge, but was unsuccessful there. The above value (let say `2.18 GiB` from the sample) is passed to one of the `ChartDonut`'s property (`title`), so I am not sure how to make it multi-line (or do any other format changes) etc.
@afreen23 , @cloudbehl , @rawagner , @yuvalgalanti  thoughts???

andybraren replied:
@aruniiird Would my `font-size` suggestion above work using the React component's `style` prop?

The backup plan might be to move the unit label into the `subTitle` prop and make more styling tweaks, but I think just adjusting the font size via the `style` prop would be simpler/better.

aruniiird replied:
> @aruniiird Would my font-size suggestion above work using the React component's style prop?

Yes, it works perfectly fine (two thumbs up)...
With @andybraren 's suggestion,

![cb_with_big_total](https://user-images.githubusercontent.com/895234/62815844-3994dd80-bb3c-11e9-879d-ae0ae9099612.png)

A few changes which would help here are,

a. `comma` being removed from `humanizedData.value` (@rawagner 's PR)...
b. if `EiB` is added to `humanizeBinaryBytes` (please see above comments and a PR is already up)
c. if we could use the `WithoutB` formats of `humanize` functions

Suggestions ???

aruniiird replied:
@andybraren , on further test, the style-fontSize combination is messing up the `tooltip`. So I am not sending a PR with those changes.

andybraren replied:
So close! Really weird that the tooltip would be affected by changing the fontSize.

This is a long shot, but would `1.2em` without the `r` work?

I think we'd prefer to keep `comma` and use `XiB` (if #2296 is approved, I could use help there). `b` sounds good though.

aruniiird replied:
I tried something like this, [ just to make sure everything written as expected ]...

```typescript
return (
  <div className="nb-capacity-card__chart-body">
    <div className="noobaa-capacity-card__item">
      <ChartDonut
        width={275}
        height={150}
        labelPosition="centroid"
        legendPosition="right"
        legendData={metricLegend}
        legendOrientation="vertical"
        data={metricData}
        labels={(datum) => `${datum.x}: ${humanizePercentage(totalUsage ? datum.y : 0).string}`}
        subTitlePosition="bottom"
        themeColor={ChartThemeColor.multi}
        themeVariant={ChartThemeVariant.light}
        title={totalHumanized.string}
        style={{ labels: { fontSize: '1em' } }}
      />
    </div>
  </div>
);
```

![image](https://user-images.githubusercontent.com/895234/62835395-b2e41b80-bc75-11e9-9df3-28189ee249a0.png)

@andybraren , any idea...

aruniiird replied:
> Here's what I see when I test something close to 1,024, at which point the unit label should increase to the next level. Seems fine if we can somehow get the fontSize to work without affecting the tooltip.

That will work even without the `style` change (it will be a bit bulkier and to the brim)...

![image](https://user-images.githubusercontent.com/895234/62835661-a6ad8d80-bc78-11e9-81b7-96b5a5400834.png)

andybraren replied:
> style={{ labels: { fontSize: '1.2em' } }}

My limited knowledge of React is failing me here 😞. Can the 1.2em be applied to something other than `labels` (plural)? Maybe `title`?

I'm not really sure why the black tooltip would be small in your earlier screenshot. If your most recent screenshot is the closest we can get then I'm happy enough for now 🙂.

cloudbehl commented on frontend/packages/noobaa-storage-plugin/src/components/capacity-card/capacity-card-body.tsx:
```diff
@@ -1,72 +1,54 @@
-};
+import {
+  LoadingInline,
+  humanizeBinaryBytesWithoutB,
```
```suggestion
  humanizeBinaryBytes,
```

aruniiird replied:
Done...

cloudbehl commented:
/hold cancel

cloudbehl commented on frontend/packages/noobaa-storage-plugin/src/components/capacity-card/capacity-card-body.tsx:
```diff
@@ -75,32 +57,31 @@ export const CapacityCardBody: React.FC<CapacityCardBodyProps> = ({ isLoading, m
+          legendOrientation="vertical"
+          data={metricData}
+          labels={(datum) => `${datum.x}: ${humanizePercentage(totalUsage ? datum.y : 0).string}`}
+          subTitlePosition="bottom"
```
are we using the subTitle?

aruniiird replied:
Acknowledeged, removed it...

cloudbehl commented:
/approve
/lgtm

@aruniiird Thanks for making the changes.
